File: sdks/dotnet/src/Mobiscroll.Connect/Resources/Webhooks.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using Mobiscroll.Connect.Models;

namespace Mobiscroll.Connect.Resources;

/// <summary>Subscribe to and unsubscribe from calendar change notifications.</summary>
public sealed class Webhooks
{
    private readonly ApiClient _client;

    internal Webhooks(ApiClient client)
    {
        _client = client;
    }

    /// <summary>
    /// Subscribe to webhook notifications for a calendar. The server auto-generates
    /// <see cref="WebhookSubscribeData.ChannelId"/> when it is omitted.
    /// </summary>
    public async Task<WebhookSubscribeResponse> SubscribeWebhookAsync(WebhookSubscribeData data, CancellationToken ct = default)
    {
        if (data is null)
        {
            throw new ArgumentNullException(nameof(data));
        }
        var result = await _client.PostJsonAsync<WebhookSubscribeResponse>("/subscribe-webhook", data, ct)
            .ConfigureAwait(false);
        return result ?? throw new InvalidOperationException("Webhook subscription returned no body");
    }

    /// <summary>
    /// Unsubscribe from webhook notifications for a calendar. Returns <c>Success: true</c>
    /// even when the provider-side subscription had already expired — check
    /// <see cref="WebhookUnsubscribeResponse.Message"/> for details in that case; treat the
    /// response as final regardless.
    /// </summary>
    public async Task<WebhookUnsubscribeResponse> UnsubscribeWebhookAsync(WebhookUnsubscribeData data, CancellationToken ct = default)
    {
        if (data is null)
        {
            throw new ArgumentNullException(nameof(data));
        }
        var result = await _client.PostJsonAsync<WebhookUnsubscribeResponse>("/unsubscribe-webhook", data, ct)
            .ConfigureAwait(false);
        return result ?? throw new InvalidOperationException("Webhook unsubscription returned no body");
    }
}

File: sdks/java/connect-sdk/src/main/java/com/mobiscroll/connect/resources/Webhooks.java
package com.mobiscroll.connect.resources;

import com.fasterxml.jackson.core.type.TypeReference;
import com.mobiscroll.connect.ApiClient;
import com.mobiscroll.connect.models.WebhookSubscribeParams;
import com.mobiscroll.connect.models.WebhookSubscribeResponse;
import com.mobiscroll.connect.models.WebhookUnsubscribeParams;
import com.mobiscroll.connect.models.WebhookUnsubscribeResponse;

/** Webhooks resource: subscribe/unsubscribe calendar change notifications. */
public final class Webhooks {

    private final ApiClient api;

    public Webhooks(ApiClient api) {
        this.api = api;
    }

    /** Subscribe to change notifications for a calendar. */
    public WebhookSubscribeResponse subscribeWebhook(WebhookSubscribeParams params) {
        return api.post("/subscribe-webhook", params, new TypeReference<WebhookSubscribeResponse>() {});
    }

    /**
     * Unsubscribe a previously created webhook channel. A {@code 200} response is final: on a
     * provider-side unsubscribe failure (e.g. an already-expired subscription) the server still
     * returns {@code success: true} with an explanatory message, after removing the local mapping.
     */
    public WebhookUnsubscribeResponse unsubscribeWebhook(WebhookUnsubscribeParams params) {
        return api.post("/unsubscribe-webhook", params, new TypeReference<WebhookUnsubscribeResponse>() {});
    }
}
